Add DivRemGeneric trait

0xpantera opened the pull request:
# Add `DivRemGeneric<T, U>` (heterogeneous div-mod) & keep `DivRem<T>` working

This PR implements the **generic, asymmetric** division-with-remainder trait that had
been marked *TODO* in `core::traits`.  
It also supplies a zero-cost compatibility bridge so **all existing code that uses the
legacy `DivRem<T>` continues to compile unchanged**.

---

## Why introduce `DivRemGeneric`?

|                     | `DivRem<T>` (before) | `DivRemGeneric<T, U>` (now) |
|---------------------|----------------------|-----------------------------|
| Operand types       | symmetric (`T / T`)  | fully generic (`T / U`)     |
| Return types        | always `(T, T)`      | associated: `(Quotient, Remainder)` |
| Expressiveness      | cannot describe `u256 / u128` | describes any integral mix that has an algorithm |
| Implementations     | hand-written per integer | keep old impls; add only the mixed-width ones you need |

### Benefits over legacy `DivRem<T>`

1. **Mixed-width math without ad-hoc helpers**  
   `DivRemGeneric<u256, u128>` lets us express `u256 / u128` directly, instead of exposing `u256_safe_div_rem` and friends.

2. **Accurate associated types**  
   Call-sites get the exact `Quotient` / `Remainder` types from the trait, enabling
   generic algorithms (e.g. `wide_div_round`) to compile for any `(T,U)` pair.

4. **No breaking changes today**  
   The bridge keeps `DivRem<T>` working, so downstream crates can migrate at their
   own pace while getting deprecation nudges.

---

## What’s in the patch?

| File | Key additions |
|------|---------------|
| `corelib/src/traits.cairo` | • Definition of `trait DivRemGeneric<T, U>`<br>• `#[deprecated] trait DivRem<T>` (unchanged API)<br>• Internal module `_divrem_bridge` that automatically provides `DivRemGeneric<T,T>` for every pre-existing `DivRem<T>` impl (u8, u16, u32, u64, u128, u256). |
| `corelib/src/integer.cairo` | No behaviour change – existing **29** symmetric impls left intact. |
| `corelib/src/test/divrem_generic_test.cairo` | New test-suite exercising:<br>  • legacy symmetric path (`u32 / u32`)<br>  • generic asymmetric path (`u256 / u128`). |
| `corelib/src/lib.cairo` | `#[feature("generic-divrem")]` gate added. |

---

## Migration / Back-compat

* **No user-visible breaking change.**  
  Code that calls `DivRem::<T>::div_rem` compiles exactly as before (now routed
  through the bridge).
* New code can opt-in to the generic form by importing `core::traits::DivRemGeneric`.
* `DivRem<T>` is soft-deprecated with the attribute:

```cairo
#[deprecated(feature: "generic-divrem", note: "Use `DivRemGeneric`.")]
```
The warning only appears when a crate doesn’t enable the new feature flag,
so corelib itself builds cleanly.

reviewable-StarkWare commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/starkware-libs/cairo/7722)
<!-- Reviewable:end -->

orizi reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/7722)* status: 0 of 7 files reviewed, 1 unresolved discussion

___
*[`corelib/src/traits.cairo` line 367 at r1](https://reviewable.io/reviews/starkware-libs/cairo/7722#-OPdZtNTEpOWp4fa6CWj:-OPdZtNj2ubgPRDz5YLn:bmtt2ge) ([raw file](https://github.com/starkware-libs/cairo/blob/254831cda8376c11e07c8a7308e4b6f74d892b0b/corelib/src/traits.cairo#L367)):*
> ```Cairo
> }
> 
> /// Performs truncated division **and** remainder.
> ```

new traits are not added here - adding traits here is deprecated.

this should probably go to `num/traits/ops` and should probably additionally lose the `Generic` in the name.


<!-- Sent from Reviewable.io -->

0xpantera commented:
Thanks for the feedback! Will move to 'num/traits/ops'. For future reference, where could I have found that adding traits in 'corelib/src/traits' was deprecated?

orizi reviewed (commented):
it was just an internal decision - you can just see all new traits were added elsewhere.

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/7722#-:-OPdfLDH6dqPKeS3cb6y:b-jalm1z)* status: 0 of 7 files reviewed, 1 unresolved discussion (waiting on @0xpantera)



<!-- Sent from Reviewable.io -->

orizi reviewed (changes requested):
> Reviewed 1 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/7722)* status: 1 of 11 files reviewed, 3 unresolved discussions (waiting on @0xpantera)

___
*[`corelib/src/traits.cairo` line 367 at r2](https://reviewable.io/reviews/starkware-libs/cairo/7722#-OPimzAfDtDjgLwOAz0a:-OPimzAfDtDjgLwOAz0b:bbrj2sk) ([raw file](https://github.com/starkware-libs/cairo/blob/ed47382d59a41886c6310f97646fbe5f41f787f0/corelib/src/traits.cairo#L367)):*
> ```Cairo
> }
> 
> /// Legacy trait for division with remainder.
> ```

revert the doc changes - this trait is by no means deprecated.
we must make sure the change fully work before such a thing.
___
*[`corelib/src/num/traits/ops/divrem.cairo` line 74 at r2](https://reviewable.io/reviews/starkware-libs/cairo/7722#-OPinejMDAA57WWzuThx:-OPinejMDAA57WWzuThy:bqyzxg2) ([raw file](https://github.com/starkware-libs/cairo/blob/ed47382d59a41886c6310f97646fbe5f41f787f0/corelib/src/num/traits/ops/divrem.cairo#L74)):*
> ```Cairo
>     pub impl BridgeU256 = Bridge<u256>;
> }
> use _divrem_bridge::*;
> ```

1. starting with `_` means unused - this is just private.
2. Either impl for all - and therefore no inner module is required, or impl specifically - and then the impl aliases should be out of the module - the `use *` did both.



_Suggestion:_
```Cairo
mod by_divrem_legacy {
    /// Generic adapter: if the old symmetric `DivRem<T>` exists,
    /// provide the corresponding `DivRemGeneric<T,T>` implementation.
    impl Impl<T, +crate::traits::DivRem<T>> of super::DivRem<T, T> {
        type Quotient = T;
        type Remainder = T;
        fn div_rem(lhs: T, rhs: NonZero<T>) -> (T, T) {
            crate::traits::DivRem::<T>::div_rem(lhs, rhs)
        }
    }
}
// Instantiate the generic adapter for every concrete integer type
// that already has a symmetric `DivRem` implementation.
pub impl DivRemU8 = by_divrem_legacy::Impl<u8>;
pub impl BridgeU16 = by_divrem_legacy::Impl<u16>;
pub impl BridgeU32 = by_divrem_legacy::Impl<u32>;
pub impl BridgeU64 = by_divrem_legacy::Impl<u64>;
pub impl BridgeU128 = by_divrem_legacy::Impl<u128>;
pub impl BridgeU256 = by_divrem_legacy::Impl<u256>;
```


<!-- Sent from Reviewable.io -->

0xpantera commented:
@orizi Appreciate the feedback. I have reverted the doc changes and fixed the legacy bridge implementation.

orizi reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/7722)* status: 1 of 25 files reviewed, 4 unresolved discussions (waiting on @0xpantera)

___
*[`-- commits` line 3 at r3](https://reviewable.io/reviews/starkware-libs/cairo/7722#-OPjgMgNEgyxOp8DbkOB:-OPjgMgNEgyxOp8DbkOC:bdkhkpv):*
fix your rebase - it now contains random stuff.


<!-- Sent from Reviewable.io -->

0xpantera commented:
Rebased onto current main, resolved the lowering conflicts by keeping
upstream changes, and rewrote the bridge as requested.
All tests green again. Thanks for your patience.

orizi reviewed (changes requested):
> Reviewed 2 of 7 files at r2, 1 of 21 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/7722)* status: 4 of 25 files reviewed, 3 unresolved discussions (waiting on @0xpantera)

___
*[`corelib/src/traits.cairo` line 380 at r4](https://reviewable.io/reviews/starkware-libs/cairo/7722#-OPy7MfyA9BRTm3BvwJ4:-OPy7MfyA9BRTm3BvwJ5:bu4ayi3) ([raw file](https://github.com/starkware-libs/cairo/blob/e557dcc8305e21b361ebec1c31567fc9fba026e3/corelib/src/traits.cairo#L380)):*
> ```Cairo
> /// ```
> pub trait DivRem<T> {
>     /// Performs the `/` and the `%` operations, returning both the quotient and remainder.
> ```

revert here as well.
___
*[`corelib/src/num/traits/ops/divrem.cairo` line 69 at r4](https://reviewable.io/reviews/starkware-libs/cairo/7722#-OPy7xrS0xJ3T6T4LCcc:-OPy7xrS0xJ3T6T4LCcd:undefined) ([raw file](https://github.com/starkware-libs/cairo/blob/a0292ef163d3a517063cdc5aa0218bf81b25a33a/corelib/src/num/traits/ops/divrem.cairo#L69)):*
> ```Cairo
> pub impl BridgeU64 = by_divrem_legacy::Bridge<u64>;
> pub impl BridgeU128 = by_divrem_legacy::Bridge<u128>;
> pub impl BridgeU256 = by_divrem_legacy::Bridge<u256>;
> ```





_Suggestion:_
```Cairo
//  Compatibility bridge:  DivRem<T>  →  DivRemGeneric<T,T>
mod by_divrem_legacy {
    /// Generic adapter: if the old symmetric `DivRem<T>` exists,
    /// provide the corresponding `DivRemGeneric<T,T>` implementation.
    pub impl Impl<T, +crate::traits::DivRem<T>> of super::DivRem<T, T> {
        type Quotient = T;
        type Remainder = T;

        fn div_rem(lhs: T, rhs: NonZero<T>) -> (T, T) {
            core::traits::DivRem::<T>::div_rem(lhs, rhs)
        }
    }
}

// Instantiate the generic adapter for every concrete integer type
// that already has a symmetric `DivRem` implementation.
impl DivRemU8 = by_divrem_legacy::Impl<u8>;
impl DivRemU16 = by_divrem_legacy::Impl<u16>;
impl DivRemU32 = by_divrem_legacy::Impl<u32>;
impl DivRemU64 = by_divrem_legacy::Impl<u64>;
impl DivRemU128 = by_divrem_legacy::Impl<u128>;
impl DivRemU256 = by_divrem_legacy::Impl<u256>;
```


<!-- Sent from Reviewable.io -->

0xpantera commented:
Thank you. Reverted doc changes and followed bridge implementation suggestion.

orizi reviewed (changes requested):
> Reviewed 1 of 7 files at r2, 2 of 21 files at r3, 12 of 13 files at r4, 3 of 3 files at r5, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/7722)* status: 24 of 25 files reviewed, 4 unresolved discussions

___
*[`corelib/src/test/num_test.cairo` line 6 at r6](https://reviewable.io/reviews/starkware-libs/cairo/7722#-OQ7HK_30XIDMuY00vAg:-OQ7HK_30XIDMuY00vAh:bfsq0lg) ([raw file](https://github.com/starkware-libs/cairo/blob/fc41188dd10a8dc62667c412c12cfeafa025ec2b/corelib/src/test/num_test.cairo#L6)):*
> ```Cairo
>     WrappingSub,
> };
> use crate::traits::DivRem as DivRemLegacy;
> ```

remove and explicitly use instead.
___
*[`corelib/src/test/num_test.cairo` line 7 at r6](https://reviewable.io/reviews/starkware-libs/cairo/7722#-OQ7H4S4EL5SJ65ww4lp:-OQ7H4S4EL5SJ65ww4lq:bfcm7qf) ([raw file](https://github.com/starkware-libs/cairo/blob/fc41188dd10a8dc62667c412c12cfeafa025ec2b/corelib/src/test/num_test.cairo#L7)):*
> ```Cairo
> };
> use crate::traits::DivRem as DivRemLegacy;
> use crate::zeroable::NonZero;
> ```

not needed.
___
*[`corelib/src/test/num_test.cairo` line 525 at r6](https://reviewable.io/reviews/starkware-libs/cairo/7722#-OQ7I0vuDU2hN5_7EtOq:-OQ7I0vuDU2hN5_7EtOr:undefined) ([raw file](https://github.com/starkware-libs/cairo/blob/fc41188dd10a8dc62667c412c12cfeafa025ec2b/corelib/src/test/num_test.cairo#L525)):*
> ```Cairo
>         (q, r_u256.try_into().unwrap())
>     }
> }
> ```





_Suggestion:_
```Cairo
#[test]
fn test_divrem() {
    assert_eq!(DivRem::<u32, u32>::div_rem(27, 4), (6, 3));
    assert_eq!(DivRem::<u256, u128>::div_rem(20, 6), (3, 2));
    // ... add more tests.
}

#[test]
fn test_divrem_legacy() {
    assert_eq!(crate::traits::DivRem::div_rem(27_u32, 4), (6, 3));
    // ... add more tests.
}

/// One-off impl for the above test.
impl U256ByU128DivRem of DivRem<u256, u128> {
    type Quotient = u256;
    type Remainder = u128;

    fn div_rem(lhs: u256, rhs: NonZero<u128>) -> (u256, u128) {
        let rhs: u128 = rhs.into();
        let rhs: u256 = rhs_u128.into();
        (lhs / rhs, (lhs % rhs).low)
    }
}
```
___
*[`corelib/src/traits.cairo` line 1136 at r6](https://reviewable.io/reviews/starkware-libs/cairo/7722#-OQ7Glxg5RU2R9BkZr6q:-OQ7Glxh5aOrHbz2R1YC:b-hay66z) ([raw file](https://github.com/starkware-libs/cairo/blob/78c3487201afb036043e085e832ccd202c69cddc/corelib/src/traits.cairo#L1136)):*
> ```Cairo
>     fn zero_default() -> T nopanic;
> }
> ```

revert.


<!-- Sent from Reviewable.io -->

orizi reviewed (changes requested):
> Reviewed 1 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/7722)* status: 24 of 25 files reviewed, 3 unresolved discussions (waiting on @0xpantera)

___
*[`corelib/src/traits.cairo` line 1136 at r6](https://reviewable.io/reviews/starkware-libs/cairo/7722#-OQ7Glxg5RU2R9BkZr6q:-OQWu3xS1oGnGF_9jWq5:b8hdbrl) ([raw file](https://github.com/starkware-libs/cairo/blob/78c3487201afb036043e085e832ccd202c69cddc/corelib/src/traits.cairo#L1136)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

revert.
</blockquote></details>

not reverted yet.
___
*[`corelib/src/test/num_test.cairo` line 486 at r7](https://reviewable.io/reviews/starkware-libs/cairo/7722#-OQWuYkv6saFXk7bPkv5:-OQWuYkw7DXTjivP7K-b:undefined) ([raw file](https://github.com/starkware-libs/cairo/blob/8336e2c0c4aa0181ff0504c42e75c7792b3e532f/corelib/src/test/num_test.cairo#L486)):*
> ```Cairo
>     );
>     assert_eq!(DivRem::<u256, u256>::div_rem(1000_u256, 33), (30_u256, 10_u256));
>     assert_eq!(DivRem::<u256, u128>::div_rem(123_u256, 10), (12_u256, 3_u128));
> ```





_Suggestion:_
```Cairo
    assert_eq!(DivRem::<u256, u256>::div_rem(1000, 33), (30, 10));
    assert_eq!(DivRem::<u256, u128>::div_rem(123, 10), (12, 3));
```
___
*[`crates/cairo-lang-lowering/src/test_data/match` line 2103 at r7](https://reviewable.io/reviews/starkware-libs/cairo/7722#-OQWuR0LDBFgX26rkUKZ:-OQWuR0LDBFgX26rkUK_:b4wcgsz) ([raw file](https://github.com/starkware-libs/cairo/blob/8336e2c0c4aa0181ff0504c42e75c7792b3e532f/crates/cairo-lang-lowering/src/test_data/match#L2103)):*
> ```
>   Return(v4)
> ```

revert file


<!-- Sent from Reviewable.io -->

orizi reviewed (changes requested):
please respond to the review on https://reviewable.io/reviews/starkware-libs/cairo/7722#-

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/7722#-:-OQsqhvI-9AI_LsOWn77:b-cw9oau)* status: 24 of 25 files reviewed, 3 unresolved discussions (waiting on @0xpantera)

___
*[`corelib/src/num/traits/ops/divrem.cairo` line 45 at r8](https://reviewable.io/reviews/starkware-libs/cairo/7722#-OQsquBI-IcMGBBI58Vk:-OQsquBJEW83nH2nH4Vi:b-dwhkcy) ([raw file](https://github.com/starkware-libs/cairo/blob/adbfc350faf958db6922dab618cd19ea87e41060/corelib/src/num/traits/ops/divrem.cairo#L45)):*
> ```Cairo
> 
>     /// Computes both `/` and `%` in a single pass.
>     fn div_rem(lhs: T, rhs: NonZero<U>) -> (Self::Quotient, Self::Remainder);
> ```

this way it would additionally support `5.div_rem(3)`.



_Suggestion:_
```Cairo
    fn div_rem(self: T, other: NonZero<U>) -> (Self::Quotient, Self::Remainder);
```


<!-- Sent from Reviewable.io -->

orizi reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/7722)* status: 22 of 25 files reviewed, 4 unresolved discussions (waiting on @0xpantera)

___
*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/7722#-OQxNP6P3nxghPFUOIMV:-OQxNP6P3nxghPFUOIMW:b-782lfd) (no related file):*
again - please respond to comment on https://reviewable.io/reviews/starkware-libs/cairo/7722#-


<!-- Sent from Reviewable.io -->

0xpantera reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/7722)* status: 22 of 25 files reviewed, 4 unresolved discussions (waiting on @orizi)

___
*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/7722#-OQxNP6P3nxghPFUOIMV:-OR0xWNz4Xfvs4qR56Rx:b-896fix) (no related file):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

again - please respond to comment on https://reviewable.io/reviews/starkware-libs/cairo/7722#-
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

orizi reviewed (changes requested):
> Reviewed 3 of 3 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/7722)* status: all files reviewed, 2 unresolved discussions (waiting on @0xpantera)

___
*[`corelib/src/traits.cairo` line 1136 at r6](https://reviewable.io/reviews/starkware-libs/cairo/7722#-OQ7Glxg5RU2R9BkZr6q:-OR51IPQ9g2z8a46GWhC:b21f6l4) ([raw file](https://github.com/starkware-libs/cairo/blob/78c3487201afb036043e085e832ccd202c69cddc/corelib/src/traits.cairo#L1136)):*
<details><summary><i>Previously, 0xpantera (Franco) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

no it isn't.
___
*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/7722#-OR51wd9AvQDM5ViaBkE:-OR51wd9AvQDM5ViaBkF:b-uuj1x0) (no related file):*
@gilbens-starkware for 2nd eye.


<!-- Sent from Reviewable.io -->

0xpantera commented:
*[`corelib/src/traits.cairo` line 1136 at r6](https://reviewable.io/reviews/starkware-libs/cairo/7722#-OQ7Glxg5RU2R9BkZr6q:-OR6vvCO3u6GRll8Nwe2:bwuhwot) ([raw file](https://github.com/starkware-libs/cairo/blob/78c3487201afb036043e085e832ccd202c69cddc/corelib/src/traits.cairo#L1136)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

no it isn't.
</blockquote></details>

Sorry if I'm misunderstanding but I reverted the changes shown and the diff in the UI no longer shows any changes. What am I missing?


<!-- Sent from Reviewable.io -->

orizi reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/7722)* status: all files reviewed, 2 unresolved discussions (waiting on @0xpantera and @gilbens-starkware)

___
*[`corelib/src/traits.cairo` line 1136 at r6](https://reviewable.io/reviews/starkware-libs/cairo/7722#-OQ7Glxg5RU2R9BkZr6q:-OR6wZGU5RO5z5_VqkU6:b-mlq0pe) ([raw file](https://github.com/starkware-libs/cairo/blob/78c3487201afb036043e085e832ccd202c69cddc/corelib/src/traits.cairo#L1136)):*
<details><summary><i>Previously, 0xpantera (Franco) wrote…</i></summary><blockquote>

Sorry if I'm misunderstanding but I reverted the changes shown and the diff in the UI no longer shows any changes. What am I missing?
</blockquote></details>

i still the removed endline in the UI.
![image.png](https://files.reviewableusercontent.io/images/reviewable/104711814/AqvZJGcA90lcTWPmTrmbZhRn/image.png)


<!-- Sent from Reviewable.io -->

0xpantera commented:
*[`corelib/src/traits.cairo` line 1136 at r6](https://reviewable.io/reviews/starkware-libs/cairo/7722#-OQ7Glxg5RU2R9BkZr6q:-OR7Ju8OEDT55lVPNB7r:b-800d16) ([raw file](https://github.com/starkware-libs/cairo/blob/78c3487201afb036043e085e832ccd202c69cddc/corelib/src/traits.cairo#L1136)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

i still the removed endline in the UI.
![image.png](https://files.reviewableusercontent.io/images/reviewable/104711814/AqvZJGcA90lcTWPmTrmbZhRn/image.png)
</blockquote></details>

Should be good now.


<!-- Sent from Reviewable.io -->

orizi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/7722#-:-OR7Xk105gojJyRFDMll:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @0xpantera and @gilbens-starkware)



<!-- Sent from Reviewable.io -->

orizi reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/7722)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @0xpantera)



<!-- Sent from Reviewable.io -->
